index.js

```diff
@@ -199,10 +203,61 @@ function session(options){
 
       setcookie(res, name, req.sessionID, secrets[0], cookie.data);
     });
+    
+    var _write = res.write;
+
+    if (saveBeforeLocationHeader) {
+      var _writeHead = res.writeHead;
+      var headWritten = false;
+      res.writeHead = function writeHead(statusCode, reason, obj) {
+        if (headWritten) return;
+        headWritten = true;
 
+        // search for location header only for 3xx status codes
+        if (statusCode < 300 || statusCode >= 400) {
+          return _writeHead.call(res, statusCode, reason, obj);
+        }
+        
+        if ('string' != typeof(reason)) {
+          obj = reason;
+          reason = undefined;
+        }
+        
+        if (obj) {
+          var keys = Object.keys(obj);
+          for (var i = 0, l = keys.length; i < l; i++) {
+            var k = keys[i];
+            if (k) res.setHeader(k, obj[k]);
+          }
+        }
+        
+        // we have a `location` header so we must buffer all writes
+        if (res.getHeader('location') != null) {
+          debug('redirect found, buffering response')
+          buffered = true;
+          bufferedChunks = new Buffer(0);
+          writeHeadLater = function() { _writeHead.call(res, statusCode, reason); };
```

Comment:
The contract with `.writeHead` is that we must call it sync right away; we cannot actually delay it's call, sadly. This is a limitation with injecting ourselves directly into the http response object without an abstraction.

The reason is what once code calls `res.writeHead()`, then Node.js guarantees that the side-effects have occurred sync, i.e the following must work, but this PR breaks it:
```js
res.writeHead(500)
assert(res.statusCode === 500)
```
```